qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,864 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=invalid-name
+
+"""Symbolic waveform module.
+
+These are pulses which are described by symbolic equations for their envelopes and for their
+parameter constraints.
+"""
+
+import functools
+from typing import Any, Dict, List, Optional, Union, Callable
+
+import numpy as np
+
+from qiskit.circuit.parameterexpression import ParameterExpression
+from qiskit.pulse.exceptions import PulseError
+from qiskit.pulse.library.pulse import Pulse
+from qiskit.pulse.library.waveform import Waveform
+from qiskit.utils import optionals as _optional
+
+if _optional.HAS_SYMENGINE:
+    import symengine as sym
+else:
+    import sympy as sym
+
+
+def _lifted_gaussian(
+    t: sym.Symbol,
+    center: Union[sym.Symbol, sym.Expr, complex],
+    t_zero: Union[sym.Symbol, sym.Expr, complex],
+    sigma: Union[sym.Symbol, sym.Expr, complex],
+) -> sym.Expr:
+    r"""Helper function that returns a lifted Gaussian symbolic equation.
+
+    For :math:`\sigma=` ``sigma`` the symbolic equation will be
+
+    .. math::
+
+        f(x) = \exp\left(-\frac12 \left(\frac{x - \mu}{\sigma}\right)^2 \right),
+
+    with the center :math:`\mu=` ``duration/2``.
+    Then, each output sample :math:`y` is modified according to:
+
+    .. math::
+
+        y \mapsto \frac{y-y^*}{1.0-y^*},
+
+    where :math:`y^*` is the value of the un-normalized Gaussian at the endpoints of the pulse.
+    This sets the endpoints to :math:`0` while preserving the amplitude at the center,
+    i.e. :math:`y` is set to :math:`1.0`.
+
+    Args:
+        t: Symbol object representing time.
+        center: Symbol or expression representing the middle point of the samples.
+        t_zero: The value of t at which the pulse is lowered to 0.
+        sigma: Symbol or expression representing Gaussian sigma.
+
+    Returns:
+        Symbolic equation.
+    """
+    gauss = sym.exp(-(((t - center) / sigma) ** 2) / 2)
+    offset = sym.exp(-(((t_zero - center) / sigma) ** 2) / 2)
+
+    return (gauss - offset) / (1 - offset)
+
+
+@functools.lru_cache(maxsize=None)
+def _is_amplitude_valid(symbolic_pulse: "SymbolicPulse") -> bool:
+    """A helper function to validate maximum amplitude limit.
+
+    Result is cached for better performance.
+
+    Args:
+        symbolic_pulse: A pulse to validate.
+
+    Returns:
+        Return True if no sample point exceeds 1.0 in absolute value.
+    """
+    try:
+        # Instantiation of Waveform does automatic amplitude validation.
+        symbolic_pulse.get_waveform()
```

Comment:
Doesn't the check that the amplitude is less 1 need to be done here?

Reply:
yeah actually it's already done in the constructor of the `Waveform`. Since this method internally instantiate the class, it will be duplicated and will be never called.